third_party/terraform/utils/retry_transport.go

```diff
@@ -0,0 +1,195 @@
+		retryPredicates: defaultErrorRetryPredicates,
+		internal:        t,
+	}
+}
+
+// Helper method to create a shallow copy of an HTTP client with a shallow-copied retryTransport
+// s.t. the base HTTP transport is the same (i.e. client connection pools are shared, retryPredicates are different)
+func ClientWithAdditionalRetries(baseClient *http.Client, baseRetryTransport *retryTransport, predicates ...RetryErrorPredicateFunc) *http.Client {
+	copied := *baseClient
+	if baseRetryTransport == nil {
+		baseRetryTransport = NewTransportWithDefaultRetries(baseClient.Transport)
+	}
+	copied.Transport = baseRetryTransport.WithAddedPredicates(predicates...)
+	return &copied
+}
+
+// Returns a shallow copy of the retry transport with additional retry
+// predicates but same wrapped http.RoundTripper
+func (t *retryTransport) WithAddedPredicates(predicates ...RetryErrorPredicateFunc) *retryTransport {
+	copyT := *t
+	copyT.retryPredicates = append(t.retryPredicates, predicates...)
+	return &copyT
+}
+
+type retryTransport struct {
+	retryPredicates []RetryErrorPredicateFunc
+	internal        http.RoundTripper
+}
+
+// RoundTrip implements the RoundTripper interface method.
+// It retries the given HTTP request based on the retry predicates
+// registered under the retryTransport.
+func (t *retryTransport) RoundTrip(req *http.Request) (resp *http.Response, respErr error) {
+	// Set timeout to default value.
+	ctx := req.Context()
+	var ccancel context.CancelFunc
+	if _, ok := ctx.Deadline(); !ok {
+		ctx, ccancel = context.WithTimeout(ctx, defaultRetryTransportTimeoutSec*time.Second)
+	}
+
+	attempts := 0
+	backoff := time.Millisecond * 500
+
+	log.Printf("[DEBUG] Retry Transport: starting RoundTrip retry loop")
+Retry:
+	for {
+		log.Printf("[DEBUG] Retry Transport: request attempt %d", attempts)
+
+		// Copy the request - we dont want to use the original request as
+		// RoundTrip contract says request body can/will be consumed
+		newRequest, copyErr := copyHttpRequest(req)
+		if copyErr != nil {
+			respErr = errwrap.Wrapf("unable to copy invalid http.Request for retry: {{err}}", copyErr)
+			break Retry
+		}
+
+		// Do the wrapped Roundtrip. This is one request in the retry loop.
+		resp, respErr = t.internal.RoundTrip(newRequest)
+		attempts++
+
+		retryErr := t.checkForRetryableError(resp, respErr)
+		if retryErr == nil {
+			log.Printf("[WARN] Retry Transport: Stopping retries, last request was successful")
+			break Retry
+		}
+		if !retryErr.Retryable {
+			log.Printf("[WARN] Retry Transport: Stopping retries, last request failed with non-retryable error: %s", retryErr.Err)
+			break Retry
+		}
+
+		log.Printf("[DEBUG] Retry Transport: Continue retry after waiting for  %s", backoff)
+		select {
+		case <-ctx.Done():
+			log.Printf("[DEBUG] Retry Transport: Stopping retries, context done: %v", ctx.Err())
+			break Retry
+		case <-time.After(backoff):
+			log.Printf("[DEBUG] Retry Transport: Finished waiting")
+			backoff *= 2
```

Comment:
Based on very quick math, I think this will be retrying at .5, 1.5, 3.5, 7.5, 15.5 and 31.5 seconds. If the default cancellation is 30s the last retry will be at 15.5s. Since that's pretty far from 30s it might be worth tweaking things around a bit. Maybe timeout at 35s? Change the level of exponential backoff? etc?